scripts.py

```diff
@@ -995,6 +1004,340 @@ class OPDSForDistributorsReaperScript(OPDSImportScript):
+    metadata and directories containing ebook and cover files.
+    """
+
+    @classmethod
+    def arg_parser(cls, _db):
+        parser = argparse.ArgumentParser()
+        parser.add_argument(
+            '--collection-name',
+            help=u'Titles will be imported into a collection with this name. The collection will be created if it does not already exist.',
+            required=True
+        )
+        parser.add_argument(
+            '--data-source-name',
+            help=u'All data associated with this import activity will be recorded as originating with this data source. The data source will be created if it does not already exist.',
+            required=True
+        )
+        parser.add_argument(
+            '--metadata-file',
+            help=u'Path to a file containing MARC metadata for every title in the collection',
+            required=True
+        )
+        parser.add_argument(
+            '--cover-directory',
+            help=u'Directory containing a full-size cover image for every title in the collection.',
+        )
+        parser.add_argument(
+            '--ebook-directory',
+            help=u'Directory containing an EPUB or PDF file for every title in the collection.',
+            required=True
+        )
+        RS = RightsStatus
+        rights_uris = ", ".join(RS.OPEN_ACCESS)
+        parser.add_argument(
+            '--rights-uri',
+            help=u"A URI explaining the rights status of the works being uploaded. Acceptable values: %s" % rights_uris,
+            required=True
+        )
+        parser.add_argument(
+            '--dry-run',
+            help=u"Show what would be imported, but don't actually do the import.",
+            action='store_true',
+        )
+        return parser
+
+    def do_run(self, cmd_args=None):
+        parser = self.arg_parser(self._db)
+        parsed = parser.parse_args(cmd_args)
+        collection_name = parsed.data_source_name
+        data_source_name = parsed.data_source_name
+        metadata_file = parsed.metadata_file
+        cover_directory = parsed.cover_directory
+        ebook_directory = parsed.ebook_directory
+        rights_uri = parsed.rights_uri
+        dry_run = parsed.dry_run
+        return self.run_with_arguments(
+            collection_name, data_source_name,
+            metadata_file, cover_directory,
+            ebook_directory, rights_uri, dry_run
+        )
+
+    def run_with_arguments(
+            self, collection_name, data_source_name,
+            metadata_file, cover_directory, ebook_directory, rights_uri,
+            dry_run
+    ):
+        if dry_run:
+            self.log.warn(
+                "This is a dry run. No files will be uploaded and nothing will change in the database."
+            )
+        collection, mirror = self.load_collection(
+            collection_name, data_source_name
+        )
+
+        if dry_run:
+            mirror = None
```

Comment:
Is it a problem if it's not a dry run but there's no mirror?

Reply:
Currently, yes, but not in the long term. Right now if there's no mirror, the cover image and ebook will be stored in the database, and the cover image won't be scaled to a thumbnail. So that's not good.

But, there's no reason the circulation manager couldn't "fulfill" an open-access book by serving a copy of the book directly from the database, or stick a thumbnail image into an OPDS feed as a `data:` URL. In this case it would be possible to serve a collection of a couple books by local authors, without having to do a full S3 integration.

So I would like to take steps towards making everything work if there's no mirror. But right now it is a problem (because you can't see the books you imported) and I'm okay with raising an exception if you try to run this script with no S3 integration.